pkg/metricshandler/metrics_handler.go

```diff
@@ -0,0 +1,244 @@
+limitations under the License.
+*/
+
+package metricshandler
+
+import (
+	"compress/gzip"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	appsv1 "k8s.io/api/apps/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog"
+	"k8s.io/kube-state-metrics/internal/store"
+	metricsstore "k8s.io/kube-state-metrics/pkg/metrics_store"
+	"k8s.io/kube-state-metrics/pkg/options"
+)
+
+type MetricsHandler struct {
+	opts               *options.Options
+	kubeClient         kubernetes.Interface
+	storeBuilder       *store.Builder
+	enableGZIPEncoding bool
+
+	cancel func()
+
+	// mtx protects stores, curShard, and curTotalShards
+	mtx            *sync.RWMutex
+	stores         []*metricsstore.MetricsStore
+	curShard       int32
+	curTotalShards int
+}
+
+func New(opts *options.Options, kubeClient kubernetes.Interface, storeBuilder *store.Builder, enableGZIPEncoding bool) *MetricsHandler {
+	return &MetricsHandler{
+		opts:               opts,
+		kubeClient:         kubeClient,
+		storeBuilder:       storeBuilder,
+		enableGZIPEncoding: enableGZIPEncoding,
+		mtx:                &sync.RWMutex{},
+	}
+}
+
+func (m *MetricsHandler) ConfigureSharding(ctx context.Context, shard int32, totalShards int) {
+	m.mtx.Lock()
+	defer m.mtx.Unlock()
+
+	if m.cancel != nil {
+		m.cancel()
+	}
+	klog.Infof("configuring sharding of this instance to be shard index %d (zero-indexed) out of %d total shards", shard, totalShards)
+	ctx, m.cancel = context.WithCancel(ctx)
+	m.storeBuilder.WithSharding(shard, totalShards)
+	m.storeBuilder.WithContext(ctx)
+	m.stores = m.storeBuilder.Build()
+	m.curShard = shard
+	m.curTotalShards = totalShards
+}
+
+func (m *MetricsHandler) Run(ctx context.Context) error {
+	autoSharding := len(m.opts.Pod) > 0 && len(m.opts.Namespace) > 0
+
+	klog.Infof("autosharding enabled: %t; pod=%v; pod_namespace=%v", autoSharding, m.opts.Pod, m.opts.Namespace)
+	if !autoSharding {
+		m.ConfigureSharding(ctx, m.opts.Shard, m.opts.TotalShards)
+		<-ctx.Done()
+		return ctx.Err()
+	}
+
+	klog.Infof("Auto detecting sharding settings.")
+	ss, err := detectStatefulSet(m.kubeClient, m.opts.Pod, m.opts.Namespace)
+	if err != nil {
+		return errors.Wrap(err, "detect StatefulSet")
+	}
+	statefulSetName := ss.Name
+
+	labelSelectorOptions := func(o *metav1.ListOptions) {
+		o.LabelSelector = fields.SelectorFromSet(fields.Set(ss.Labels)).String()
+	}
+
+	i := cache.NewSharedIndexInformer(
+		cache.NewFilteredListWatchFromClient(m.kubeClient.AppsV1().RESTClient(), "statefulsets", m.opts.Namespace, labelSelectorOptions),
+		&appsv1.StatefulSet{}, 0, cache.Indexers{cache.NamespaceIndex: cache.MetaNamespaceIndexFunc},
+	)
+	i.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(o interface{}) {
+			ss := o.(*appsv1.StatefulSet)
+			if ss.Name != statefulSetName {
```

Comment:
Let's address the code duplication in AddFunc and UpdateFunc.

Reply:
I don't think this is worth it, and the "skip" logic is slightly different for add and update, as in update we also get to compare to the previous object. I think this duplication is ok in its current extend.